protos/feast/serving/ServingService.proto

```diff
@@ -19,55 +19,140 @@ syntax = "proto3";
 package feast.serving;
 
 import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
 import "feast/types/Value.proto";
 
 option java_package = "feast.serving";
 option java_outer_classname = "ServingAPIProto";
 option go_package = "github.com/gojek/feast/protos/generated/go/feast/serving";
 
 service ServingService {
-    // Query features from Feast serving storage
-    rpc QueryFeatures (QueryFeaturesRequest) returns (QueryFeaturesResponse) {
-    };
-
     // Retrieve version information about this Feast deployment
-    rpc GetFeastServingVersion (GetFeastServingVersionRequest) returns (GetFeastServingVersionResponse);
-}
+    rpc GetFeastServingVersion (google.protobuf.Empty) returns (GetFeastServingVersionResponse);
+
+    // Get Feast serving storage type (online or offline)
+    rpc GetFeastServingType (google.protobuf.Empty) returns (GetFeastServingTypeResponse);
+
+    // Get online features from Feast serving
+    rpc GetOnlineFeatures (GetFeaturesRequest) returns (GetOnlineFeaturesResponse);
+
+    // Get offline features from Feast serving. This is an async job.
+    rpc GetOfflineFeatures (GetFeaturesRequest) returns (GetOfflineFeaturesResponse);
+
+    // Get the current status of an offline feature request job
+    rpc GetOfflineFeaturesJobStatus (OfflineFeaturesJob.GetStatusRequest) returns (OfflineFeaturesJob.GetStatusResponse);
 
-message QueryFeaturesRequest {
-    // e.g. "driver", "customer", "city".
-    string entityName = 1;
-    // List of entity ID.
-    repeated string entityId = 2;
-    // List of feature ID.
-    // feature ID is in the form of [entity_name].[feature_name]
-    // e.g: "driver.day.total_accepted_booking"
-    // all requested feature ID shall have same entity name.
-    repeated string featureId = 3;
+    // Request a signed URL where a Feast client can upload user entity data
+    rpc GetOfflineFeaturesJobUploadUrl (OfflineFeaturesJob.GetUploadUrlRequest) returns (OfflineFeaturesJob.GetUploadUrlResponse);
+
+    // Set the state of the offline feature job to complete after user entity data has been uploaded
+    rpc SetOfflineFeaturesJobUploadComplete (OfflineFeaturesJob.SetUploadCompleteRequest) returns (OfflineFeaturesJob.SetUploadCompleteResponse);
 }
 
-message QueryFeaturesResponse {
-    // Entity name of the response
-    string entityName = 1;
-    // map of entity ID and its entity's properties.
-    map<string, Entity> entities = 2;
+message GetFeaturesRequest {
+    repeated FeatureId featureIds = 1;
+    UserEntityData userEntityData = 2;
+
+    message FeatureId {
+        string featureSetName = 1;
+        string featureSetVersion = 2;
+        string featureName = 3;
+    }
+
+    message UserEntityData {
```

Comment:
UserEntity might be a word that is misconstrued if you plan on adding entities back in, imo

Reply:
Are you able to come up with a better name?